site/content/blog/2020-05-20-hands-on-iouring-go.md

```diff
@@ -0,0 +1,327 @@
+
+### Polled mode
+
+For latency sensitive applications, or ones with extremely high IOPS, letting the device driver interrupt the kernel every time data is available to be read is not efficient enough. If we have a lot of data to be read, a high rate of interrupts would actually slow down the kernel throughput for processing the events. In those situations, we actually fall back to polling the device driver. To use polling with `io_uring`, we can set the `IORING_SETUP_IOPOLL` flag in the `io_uring_setup` call and keep polling events with the `IORING_ENTER_GETEVENTS` set in the `io_uring_enter` call.
+
+But this still requires "us", the user, to make the calls. To take things up even a notch higher, `io_uring` also has a feature calling "kernel-side polling", whereby if we set the `IORING_SETUP_SQPOLL` flag in `io_uring_params`, the kernel will poll the SQ automatically to check for new entries, and consume them. This essentially means that we can keep doing all the I/O we want without performing even a _single_. _system_. _call_. This changes everything.
+
+But all of this flexibility and raw power comes at a cost. Using this API directly is non-trivial and error-prone. Since our data structures are shared between user and kernel, we need to setup memory barriers (magical compiler incantations to enforce ordering of memory operations) and other nitty-gritties to get things done properly.
+
+Fortunately, Jens Axboe, the creator of `io_uring` has created a wrapper library called `liburing` to help simplify all of this. With `liburing`, we roughly have to do this set of steps:
+
+- `io_uring_queue_(init|exit)` to setup and tear down the io_uring.
+
+- `io_uring_get_sqe` to get an SQE.
+
+- `io_uring_prep_(readv|writev|other)` to mark which syscall to use.
+
+- `io_uring_sqe_set_data` to mark the user data field.
+
+- `io_uring_(wait|peek)_cqe` to either wait for cqe or peek for it without waiting.
+
+- `io_uring_cqe_get_data` to get back the user data field.
+
+- `io_uring_cqe_seen` to mark the CQE as done.
+
+### Wrapping io_uring in Go
+
+This was a lot of theory to digest. There's even more of it which I have deliberately skipped for the sake of brevity. Now let's get back to writing some code in Go and take this for a spin. For simplicity and safety we will use the `liburing` library, which means we will need to use CGo. That's fine because this is just a toy, and the right way would be have [native support](https://github.com/golang/go/issues/31908) in the Go runtime. As a result of that, we will unfortunately have to use callbacks. In native Go, the running goroutine would be put to sleep by the runtime, and then woken up when the data would be available in the completion queue.
```

Comment:
```suggestion
This was a lot of theory to digest. There's even more of it which I have deliberately skipped for the sake of brevity. Now let's get back to writing some code in Go and take this for a spin. For simplicity and safety we will use the `liburing` library, which means we will need to use CGo. That's fine because this is just a toy, and the right way would be have [native support](https://github.com/golang/go/issues/31908) in the Go runtime. As a result of that, we will unfortunately have to use callbacks. In native Go, the running `goroutine` would be put to sleep by the runtime, and then woken up when the data would be available in the completion queue.
```

Reply:
Discussed offline. Not needed.